### PR TITLE
[MUL-298] Build repairs - Removal of deprecated EntityManager::merge() method

### PR DESCRIPTION
**Pull request by DominikWolos**

### CODE REVIEW

File: src/Component/ProductListing/Entity/DraftAttributeValue.php
Comment (DominikWolos) [inline src/Component/ProductListing/Entity/DraftAttributeValue.php:1]: This approach eliminates bug related to adding attributes to product listing. 
@senghe I assumed that PHPStan was the only purpose of this change, am I correct?

File: templates/Context/Vendor/DraftAttributes/update.html.twig
Comment (DominikWolos) [inline templates/Context/Vendor/DraftAttributes/update.html.twig:1]: Fixes bug preventing from editing attribute

File: features/admin/accepting_product_listing.feature [hunk 1]
```diff
@@ -35,3 +35,22 @@ Feature: Verifying product listing
     And I should see url "#\/admin\/product-listings\/(\d+)#"
     When I click "Accept" button
     Then there should be product with channel enabled
+
+    @ui
+    Scenario: Accept updated product listing with attributes
+    Given there is product listing enabled for channel
+    And there is a product 'green-jeans' for listing
```
Comment (senghe) [inline features/admin/accepting_product_listing.feature:42]:
```suggestion
    And there is a product 'green-jeans' attached to the product listing
```

File: features/admin/accepting_product_listing.feature [hunk 2]
```diff
@@ -35,3 +35,22 @@ Feature: Verifying product listing
...
+    And there is draft attribute with code 'attribute-1' and type 'text'
+    And there is draft attribute with code 'attribute-2' and type 'text'
+    And there is already published product with attribute 'attribute-1' with value 'value-1'
+    And product listing has attribute 'attribute-2' with value 'value-2'
+    And I am on "/admin"
```
Comment (senghe) [inline features/admin/accepting_product_listing.feature:47]: This is the "When" section I think. "When I go to "/admin"

File: features/admin/accepting_product_listing.feature [hunk 3]
```diff
@@ -35,3 +35,22 @@ Feature: Verifying product listing
...
+    And I follow "Product listings"
+    And I should see 1 product listing
+    And I should see product's listing status "Under verification"
+    And I follow "Details"
+    And I should see url "#\/admin\/product-listings\/(\d+)#"
+    When I click "Accept" button
+    And I follow "Details"
+    And I should see 'attribute-2' with value 'value-2'
+    And I should not see 'attribute-1' with value 'value-1'
```
Comment (senghe) [inline features/admin/accepting_product_listing.feature:56]: There is no "Then" section. Additionally I see you try to test two things in one test, which is bad idea.

File: features/vendor/create_product_listing_with_attributes.feature [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    And I follow "Product list"
+    And I follow "Create Product"
+    When I fill in "Code" with "productTest"
+    And I fill in "Price" with "10"
+    And I fill in "Original price" with "20"
+    And I fill in "Minimum price" with "30"
+    And I fill in "Name" with "test"
+    And I fill in "Slug" with "product"
+    And I select "extended" from "sylius_product_attribute_choice"
+    And I click "Save" button
```
Comment (senghe) [inline features/vendor/create_product_listing_with_attributes.feature:27]: There is no "Then" section.

File: src/Component/ProductListing/Entity/DraftAttributeValue.php [hunk 5]
```diff
@@ -44,10 +44,8 @@ public function setDraft(?DraftInterface $product): void
         parent::setSubject($product);
     }
 
-    public function getAttribute(): ?DraftAttributeInterface
```
Comment (senghe) [inline src/Component/ProductListing/Entity/DraftAttributeValue.php:47]: Does the method come from any Sylius' interface? If no, I would not approve this change.

**Review by senghe — CHANGES_REQUESTED** [review]

File: src/Component/ProductListing/Entity/DraftAttributeValue.php [hunk 6]
```diff
@@ -44,10 +44,8 @@ public function setDraft(?DraftInterface $product): void
         parent::setSubject($product);
     }
```
Comment (senghe) [inline src/Component/ProductListing/Entity/DraftAttributeValue.php:46]: Please leave PHPDoc with /** @return ProductDraftAttributeInterface */

File: src/Component/ProductListing/Entity/DraftAttributeValue.php (hunk 5 above)
Comment (DominikWolos) [reply src/Component/ProductListing/Entity/DraftAttributeValue.php:47]: https://github.com/BitBagCommerce/OpenMarketplace/pull/132#discussion_r1423792196 Related comment

File: features/vendor/create_product_listing.feature [hunk 7]
```diff
@@ -37,5 +37,5 @@ Feature:Creating a product listing.
     And I fill in "Minimum price" with "30"
     And I fill in "Name" with "test"
     And I fill in "Slug" with "product"
-    When I click "Save" button
+    Then I click "Save" button
     Then I should see non unique code error message
```
Comment (senghe) [inline features/vendor/create_product_listing.feature:41]:
```suggestion
    And I should see non unique code error message
```

File: features/vendor/create_product_listing.feature [hunk 8]
```diff
@@ -37,5 +37,5 @@ Feature:Creating a product listing.
     And I fill in "Minimum price" with "30"
     And I fill in "Name" with "test"
     And I fill in "Slug" with "product"
-    When I click "Save" button
+    Then I click "Save" button
```
Comment (senghe) [inline features/vendor/create_product_listing.feature:40]:
```suggestion
    And I click "Save" button
```

File: features/vendor/create_product_listing.feature [hunk 9]
```diff
@@ -37,5 +37,5 @@ Feature:Creating a product listing.
     And I fill in "Minimum price" with "30"
     And I fill in "Name" with "test"
     And I fill in "Slug" with "product"
```
Comment (senghe) [inline features/vendor/create_product_listing.feature:39]: I think from the scenario perspective all the data you set aren't so much important to have specified in the Gherkin scenario. I would call this "And I fill the data with "code0" code which is invalid". The rest of the data can be set implicitely in the behat context, even in this step.

File: src/Component/ProductListing/Entity/DraftAttributeValue.php [hunk 10]
```diff
@@ -44,9 +44,23 @@ public function setDraft(?DraftInterface $product): void
...
 
-    public function getAttribute(): ?DraftAttributeInterface
+    public function setAttribute(?AttributeInterface $attribute): void
     {
-        Assert::isInstanceOf($this->attribute, DraftAttributeInterface::class);
+        if (null !== $attribute && !$attribute instanceof DraftAttributeInterface) {
+            throw new \InvalidArgumentException(
```
Comment (senghe) [inline src/Component/ProductListing/Entity/DraftAttributeValue.php:50]: please use `Assert::isInstanceOf` :pray:

**Review by senghe — CHANGES_REQUESTED** [review]

File: features/vendor/create_product_listing.feature (hunk 9 above)
Comment (DominikWolos) [reply features/vendor/create_product_listing.feature:39]: Can't we leave it as is since all of this actually presents that product is valid?

Comment (DominikWolos) [reply features/vendor/create_product_listing.feature:39]: Initially wanted to discuss this, fixed

**Pull request merged**